File: src/main/java/weld/Platform.java
package weld;

import sun.misc.Cleaner;
import sun.misc.Unsafe;

import java.lang.reflect.Constructor;

/**
 * Platform dependent operations.
 */
class Platform {
  private static final Unsafe UNSAFE;

  static {
    Unsafe unsafe;
    try {
      Constructor<Unsafe> ctor = Unsafe.class.getDeclaredConstructor();
      ctor.setAccessible(true);
      unsafe = ctor.newInstance();
    } catch (ReflectiveOperationException e) {
      unsafe = null;
    }
    UNSAFE = unsafe;
  }

  /**
   * Register an AutoCloseable for automatic clean-up as soon as it gets garbage collected.
   */
  public static void registerForCleanUp(AutoCloseable closeable) {
    Cleaner.create(closeable, () -> {
      try {
        closeable.close();
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    });
  }

  public static long allocateMemory(long size) {
    return UNSAFE.allocateMemory(size);
  }

  public static void freeMemory(long address) {
    UNSAFE.freeMemory(address);
  }

  public static void copyMemory(long src, long dest, long length) {
    UNSAFE.copyMemory(null, src, null, dest, length);
  }

  public static byte getByte(long address) {
    return UNSAFE.getByte(address);
  }

  public static void putByte(long address, byte value) {
    UNSAFE.putByte(address, value);
  }

  public static void putBytes(long address, byte[] values) {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_BOOLEAN_BASE_OFFSET, null, address, values.length);
  }

  public static int getInt(long address) {
    return UNSAFE.getInt(address);
  }

  public static void putInt(long address, int value) {
    UNSAFE.putInt(address, value);
  }

  public static void putInts(long address, int[] values) {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_INT_BASE_OFFSET, null, address, values.length * 4);
  }

  public static float getFloat(long address) {
    return UNSAFE.getFloat(address);
  }

  public static void putFloat(long address, float value) {
    UNSAFE.putFloat(address, value);
  }

  public static void putFloats(long address, float[] values) {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_FLOAT_BASE_OFFSET, null, address, values.length * 4);
  }

  public static long getLong(long address) {
    return UNSAFE.getLong(address);
  }

  public static void putLong(long address, long value) {
    UNSAFE.putLong(address, value);
  }

  public static void putLongs(long address, long[] values) {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_LONG_BASE_OFFSET, null, address, values.length * 8);
  }

  public static double getDouble(long address) {
    return UNSAFE.getDouble(address);
  }

  public static void putDouble(long address, double value) {
    UNSAFE.putDouble(address, value);
  }

  public static void putDoubles(long address, double[] values) {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_DOUBLE_BASE_OFFSET, null, address, values.length * 8);
  }
}

File: src/main/scala/weld/Platform.scala
package weld

import sun.misc.{Cleaner, Unsafe}

/**
 * Platform dependent operations.
 */
object Platform {
  private val UNSAFE = {
    val ctor = classOf[Unsafe].getDeclaredConstructor()
    ctor.setAccessible(true)
    ctor.newInstance()
  }

  /**
   * Register an AutoCloseable for automatic clean-up as soon as it gets garbage collected.
   */
  def registerForCleanUp(closeable: AutoCloseable): Unit = {
    Cleaner.create(closeable, new Runnable {
      override def run(): Unit = {
        try closeable.close() catch {
          case e: Exception =>
            throw new RuntimeException(e)
        }
      }
    })
  }

  def allocateMemory(size: Long): Long = UNSAFE.allocateMemory(size)

  def freeMemory(address: Long): Unit = {
    UNSAFE.freeMemory(address)
  }

  def copyMemory(src: Long, dest: Long, length: Long): Unit = {
    UNSAFE.copyMemory(null, src, null, dest, length)
  }

  def getByte(address: Long): Byte = UNSAFE.getByte(address)

  def putByte(address: Long, value: Byte): Unit = {
    UNSAFE.putByte(address, value)
  }

  def putBytes(address: Long, values: Array[Byte]): Unit = {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_BOOLEAN_BASE_OFFSET, null, address, values.length)
  }

  def getInt(address: Long): Int = UNSAFE.getInt(address)

  def putInt(address: Long, value: Int): Unit = {
    UNSAFE.putInt(address, value)
  }

  def putInts(address: Long, values: Array[Int]): Unit = {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_INT_BASE_OFFSET, null, address, values.length * 4)
  }

  def getFloat(address: Long): Float = UNSAFE.getFloat(address)

  def putFloat(address: Long, value: Float): Unit = {
    UNSAFE.putFloat(address, value)
  }

  def putFloats(address: Long, values: Array[Float]): Unit = {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_FLOAT_BASE_OFFSET, null, address, values.length * 4)
  }

  def getLong(address: Long): Long = UNSAFE.getLong(address)

  def putLong(address: Long, value: Long): Unit = {
    UNSAFE.putLong(address, value)
  }

  def putLongs(address: Long, values: Array[Long]): Unit = {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_LONG_BASE_OFFSET, null, address, values.length * 8)
  }

  def getDouble(address: Long): Double = UNSAFE.getDouble(address)

  def putDouble(address: Long, value: Double): Unit = {
    UNSAFE.putDouble(address, value)
  }

  def putDoubles(address: Long, values: Array[Double]): Unit = {
    UNSAFE.copyMemory(values, Unsafe.ARRAY_DOUBLE_BASE_OFFSET, null, address, values.length * 8)
  }
}
